### DISCUSSION TITLE
INITIAL_MEMORY, MAXIMUM_MEMORY cannot be set dynamically

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#21931] [Q&A]
INITIAL_MEMORY, MAXIMUM_MEMORY cannot be set dynamically  like PTHREAD_POOL_SIZE for example.
it is crucial for support in different Platforms  (e.g. iOS vs Windows)

**Comment by sbc100** [comment]
Normally with emscripten the memory is part of the wasm file and this is exported to JS, so the intial and max values are baked into the binary.

If you want to control them dynamically you can use the `-sIMPORTED_MEMORY` setting and then defined you own memory dynamiclally on the `Module` object. e.g.

```
Module.wasmMemory = new WebAssembly.Memory(...).
```

BTW there some reason is not enough to just always set a low `INITIAL_MEMORY` (small enough for your smallest device) and a high `MAXIMUM_MEMORY` (big enough for your biggest device)?

**Comment by aharondavid** [comment]
My goal is to create one binary wasm that fit to Windows (4g) and iOS device (less than 1G) as well.
In some iOS devices, MAXIMUM_MEMORY cannot be greater than 1G . the following Emscripten.JS code  is crash in those devices:
```
wasmMemory = new WebAssembly.Memory({
   "initial": INITIAL_MEMORY / 65536,
   "maximum": 2147483648 / 65536,
   "shared": true
  });
```


About the  IMPORTED_MEMORY flag, it is  not clear who to work with this flag. 
I don't see any change in Emscripten.js with or without this flag.

**Reply by sbc100** [reply]
I was not aware of the this limitation on `MAXIMUM_MEMORY`.  It seems like the module should only fail to load if `INITIAL_MEMORY` is too big... so that sounds like a bug in iOS.

You can use  `IMPORTED_MEMORY` in combination with `Module['wasmMemory']`.   For example you could add the following to your `--pre-js` file: https://github.com/emscripten-core/emscripten/blob/2bc5e3156f07e603bc4f3580cf84c038ea99b2df/test/core/test_module_wasm_memory.js#L8

See this test for a full example: https://github.com/emscripten-core/emscripten/blob/2bc5e3156f07e603bc4f3580cf84c038ea99b2df/test/test_core.py#L2166-L2174

**Comment by aharondavid** [comment]
- It seems that `USE_PTHREADS  `disable the `IMPORTED_MEMORY` flag:

   Steps to reproduce (emcc 3.1.55):
  
  1.
```
  create test.cpp:
  int main() {
      //myFunction();
      return 0;
  };
```

  2. `emcc test.cpp -o output1.html -s WASM=1 -s USE_PTHREADS  -s IMPORTED_MEMORY=1`
  3. `emcc test.cpp -o output2.html -s WASM=1 -s USE_PTHREADS  `
  4. result: `output1.js` is equal to `output2.js`
  only without `USE_PTHREADS`  the `IMPORTED_MEMORY` is applied. 


- What about  `MAXIMUM_MEMORY=4gb` (4gb has special treatments of generating to the output Emscripten JS file).  
is 4gb can be set dynamically ?

**Reply by sbc100** [reply]
The code for handling memories larger than 2gb is baked into the output, but only when `MAXIMUM_MEMORY` is >2gb.   Sadly there is no way to opt into this other than setting `MAXIMUM_MEMORY` > 2gb.

**Reply by sbc100** [reply]
Its actually the other way around: `USE_PTHREADS` forces the use of `IMPORTED_MEMORY`.   This is a necessary detail of how pthreads are implemented.

**Comment by aharondavid** [comment]
Actually, MAXIMUM_MEMORY  doesn't need be allocated dynamically. 
This code in `pre-js` is works for all OS when MAXIMUM_MEMORY=4g:

```
////////////////////////////////////////////////////////////////////
// Fix iOS (17.0) bug:  WebAssembly.Memory.maximum cannot be greater than 1gb
/////////////////////////////////////////////////////
function GetMaximumMemory(){

	if (isMobileDevice()) {
		if (iOS())
			return 700*1024*1024;//700mb
	}
	return 4* 1024*1024*1024;//4gb
}


if (Module["wasmMemory"] == undefined)
	Module['wasmMemory'] = new WebAssembly.Memory({ 'initial': 300 * 1024 * 1024 / 65536, 'maximum': GetMaximumMemory() / 65536, "shared": true });

////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////
```
